### PR TITLE
Android: Migrate auto-complete to Flow and Coroutines

### PR DESCRIPTION
**Pull request by anikiki**
Task/Issue URL: https://app.asana.com/0/1200581511062568/1208040472309277/f

### Description
Removed rxjava from autosuggest.

### Steps to test this PR

Smoke test the autosuggest. There should be no user facing changes.

This is the final PR. See scenarios and screenshots in https://app.asana.com/0/1200581511062568/1208264037618741/f

### NO UI changes

### CODE REVIEW

**Comment by anikiki** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5123?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5123** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5123?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#5105** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5105?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5065** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5065?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5064** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5064?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5063** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5063?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5042** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5042?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5019** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5019?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5018** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5018?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @anikiki and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt [hunk 1]
```diff
@@ -134,33 +138,30 @@ class AutoCompleteApi @Inject constructor(
...
-                    getHistoryResults(query),
-                ) { bookmarksAndFavoritesAndTabs, historyItems ->
-                    val searchHistory = historyItems.filter { it.suggestion is AutoCompleteHistorySearchSuggestion }
-                    val navigationHistory = historyItems
-                        .filter { it.suggestion is AutoCompleteHistorySuggestion } as List<RankedSuggestion<AutoCompleteHistorySuggestion>>
-                    (removeDuplicates(navigationHistory, bookmarksAndFavoritesAndTabs) + searchHistory)
-                        .sortedByDescending { it.score }
-                        .map { it.suggestion }
-                }
+        val savedSites = getAutoCompleteBookmarkResults(query)
```
Comment (CrisBarreiro) [inline app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt:147]: For simplicity, can we just combine all flows in one single operation?

File: app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt [hunk 2]
```diff
@@ -134,33 +138,30 @@ class AutoCompleteApi @Inject constructor(
...
+                (removeDuplicates(navigationHistory, bookmarksAndFavoritesAndTabs) + searchHistory)
+                    .sortedByDescending { it.score }
+                    .map { it.suggestion }
+            }.combine(getAutoCompleteSearchResults(query)) { bookmarksAndFavoritesAndTabsAndHistory, searchResults ->
+                Pair(bookmarksAndFavoritesAndTabsAndHistory, searchResults)
+            }
 
-        return savedSitesObservable.zipWith(getAutoCompleteSearchResults(query)) { bookmarksAndTabsAndHistory, searchResults ->
-            val topHits = (searchResults + bookmarksAndTabsAndHistory).filter {
+        savedSites.collect { (bookmarksAndFavoritesAndTabsAndHistory, searchResults) ->
```
Comment (CrisBarreiro) [inline app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt:163]: Can't we just `map` instead of `collect` and `emit`?

File: app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt [hunk 3]
```diff
@@ -468,3 +466,15 @@ class AutoCompleteApi @Inject constructor(
         val score: Int = DEFAULT_SCORE,
     )
 }
+
+@VisibleForTesting
```
Comment (CrisBarreiro) [inline app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt:470]: Do we need the annotation since this is internal?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 4]
```diff
@@ -1681,6 +1681,9 @@ class BrowserTabFragment :
                 duckPlayerScripts.sendSubscriptionEvent(it.duckPlayerData)
             }
             is Command.SwitchToTab -> {
+                binding.focusedView.gone()
```
Comment (CrisBarreiro) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:1684]: Curious why we need this, since it doesn't seem to be related to the migration

**Review by CrisBarreiro — APPROVED** [review]
Looks good and works as expected. Left a coupe of small comments, but great work! 🎉 

Now that we aren't using Rx for autocomplete anymore, can we remove the rxjava dependencies somewhere? I believe we should be able to remove them from at least the history modules

File: app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt [hunk 5]
```diff
@@ -1465,7 +1463,8 @@ class BrowserTabViewModelTest {
             whenever(mockAutoCompleteScorer.score("title", "https://foo.com".toUri(), 1, "title")).thenReturn(1)
             whenever(mockUserStageStore.getUserAppStage()).thenReturn(ESTABLISHED)
 
-            testee.autoCompletePublishSubject.accept("title")
+            testee.autoCompleteStateFlow.value = "title"
+            delay(500)
```
Comment (Marchuck) [inline app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:1467]: shouldn't be `advanceTimeBy(500)` ?

File: app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt (hunk 3 above)
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt:470]: The annotation is there to mark the fact that this function should be private, but the visibility was changed for testing purposes. I didn't add what the visibility should have been. By default, if not mentioned,  is private (as in this case).

`@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)`

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt (hunk 4 above)
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:1684]: This code was added to cover the scenario where the `Switch to tab` suggestion matches the current tab. In this case we need to close the suggestions and have the UI in a state similar as when opening a different tab.

File: app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt (hunk 1 above)
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/autocomplete/api/AutoComplete.kt]: Thanks for the suggestion :+1: Updated.

File: app/src/test/java/com/duckduckgo/app/systemsearch/SystemSearchViewModelTest.kt [hunk 6]
```diff
@@ -528,15 +539,15 @@ class SystemSearchViewModelTest {
         val suggestion = AutoCompleteHistorySuggestion(phrase = "phrase", title = "title", url = "url", isAllowedInTopHits = false)
         val omnibarText = "foo"
 
-        val testObserver = TestObserver.create<String>()
-        testee.resultsPublishSubject.subscribe(testObserver)
+//        val testObserver = TestObserver.create<String>()
+//        testee.resultsStateFlow.subscribe(testObserver)
```
Comment (joshliebe) [inline app/src/test/java/com/duckduckgo/app/systemsearch/SystemSearchViewModelTest.kt:543]: You can delete the commented-out lines in this file

Comment (anikiki) [reply app/src/test/java/com/duckduckgo/app/systemsearch/SystemSearchViewModelTest.kt]: Thanks! Removed :)
